### PR TITLE
[FIX] fixes for parsing modification information in pepXML files

### PR DESCRIPTION
**Pull request by hendrikweisser**
Note to reviewers: The part that was deleted in PepXMLFile.cpp (`if (!desc.empty()) ...`) is repeated later (https://github.com/OpenMS/OpenMS/compare/develop...hendrikweisser:fix-pepxml-mods?expand=1#diff-0ce89f78310b558a3cb2058c1337de32L1269). It was redundant after I fixed the `desc = ...` line.

### CODE REVIEW

File: src/openms/source/CHEMISTRY/ModificationsDB.cpp [hunk 1]
```diff
@@ -110,7 +110,15 @@ namespace OpenMS
   const ResidueModification& ModificationsDB::getModification(const String& mod_name, const String& residue, ResidueModification::TermSpecificity term_spec) const
   {
     set<const ResidueModification*> mods;
-    searchModifications(mods, mod_name, residue, term_spec);
+    // if residue is specified, try residue-specific search first to avoid
+    // ambiguities (e.g. "Carbamidomethyl (N-term)"/"Carbamidomethyl (C)"):
+    if (!residue.empty() &&
+        (term_spec == ResidueModification::NUMBER_OF_TERM_SPECIFICITY))
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/ModificationsDB.cpp]: Out of curiosity: I never saw that term_spec == ResidueModification::NUMBER_OF_TERM_SPECIFICITY. When does this actually happen?

**Comment by timosachsenberg** [comment]
@hendrikweisser  do you have some additional information, e.g., which types of pepXML do trigger the problem?

File: src/openms/source/CHEMISTRY/ModificationsDB.cpp (hunk 1 above)
Comment (hendrikweisser) [inline src/openms/source/CHEMISTRY/ModificationsDB.cpp]: That's the default, i.e. for when you don't want to set any particular specificity. It's unfortunate that default values for function parameters are specified in the header files, not (also) in the implementation.

**Comment by hendrikweisser** [comment]
I tried to convert iProphet results from the iPRG-2015 study (data: ftp://ftp.peptideatlas.org/distro/id/pepXML/). The pepXMLs are above 80 MB each, so would need to be reduced to make test cases.

The modifications were specified there as follows:
```
<aminoacid_modification aminoacid="C" massdiff="57.021000000000001" mass="160.03019" variable="N" description="Carbamidomethyl"/>
<aminoacid_modification aminoacid="M" massdiff="15.995000000000001" mass="147.03549" variable="Y" description="Oxidation"/>
<aminoacid_modification aminoacid="C" massdiff="-17.026" mass="143.00419" variable="Y" peptide_terminus="N" description="Ammonia-loss"/>
<aminoacid_modification aminoacid="E" massdiff="-18.010000000000002" mass="111.03259" variable="Y" peptide_terminus="N" description="Glu-&gt;pyro-Glu"/>
<aminoacid_modification aminoacid="Q" massdiff="-17.026" mass="111.03258" variable="Y" peptide_terminus="N" description="Ammonia-loss"/>
<terminal_modification terminus="n" massdiff="42.011000000000003" mass="43.01894" variable="Y" protein_terminus="Y" description="Acetyl"/>
```

Before the changes, I got various errors about unrecognized modifications in the pepXML and an exception from `ModificationsDB::getModification`. I fixed those problems, so now the only complaint is about "Ammonia-loss (Q)", which is indeed not allowed according to Unimod.

**Comment by hroest** [comment]
@hendrikweisser could you maybe add the offending lines to the current testing pep.xml so that we have those as regression tests?

**Comment by hendrikweisser** [comment]
Okay, I'll add a test of some kind. (Might take me a bit to get to it.)

**Comment by hendrikweisser** [comment]
I've added a test case, but that actually contains one search hit (the last one) where our conversion is incorrect (see https://github.com/OpenMS/OpenMS/pull/2254/files#diff-3c2a70c37d0946296727110d0aa2383dR50). There are two problems:
1. The modification that gets assigned ("Pyro-carbamidomethyl") doesn't match any of the modifications in the search parameters.
2. We incorrectly add a "Carbamidomethyl" to the already modified residue.

There are multiple factors that cause this:
- When parsing pepXML, we parse modification definitions ("aminoacid_modification" and "terminal_modification") by name (preferentially). However, for occurrences of modifications in the search hits, we perform look-up by mass value ("mod_aminoacid_mass"), without consideration of which allowed modifications were defined for the search.
- At least in this example, "Carbamidomethyl (C)" was defined as a fixed modification, then "Ammonia-loss (N-term C)" was defined as a variable modification *of the carbamidomethylated C*. (See https://github.com/OpenMS/OpenMS/pull/2254/files#diff-3182549c42e42aca0623bcd56b440d66R65 - the "mass" and "massdiff" values add up to the mass of the modified C.) Both modifications together are equivalent to "Pyro-carbamidomethyl (N-term C)", which is what our parser finds.
- Terminal modifications with residue specificity (e.g. "N-term C") are assigned to the peptide terminus by us, not to the residue (see https://github.com/OpenMS/OpenMS/issues/2085#issuecomment-238224076). Due to this, the pepXML parser adds the "Carbamidomethyl" to the C residue, since that was defined as a fixed modification. Unfortunately this combination ("(Pyro-carbamidomethyl)C(Carbamidomethyl)...") is incorrect; either "Pyro-carbamidomethyl)C..." or "(Ammonia-loss)C(Carbamidomethyl)..." would have been fine.

Unfortunately fixing this is currently too involved for me. Our code for parsing modifications in pepXML is really complex. Also, it may be necessary to rethink our handling of terminal mods with residue specificity - perhaps they do need to go on the residue, not the terminus -, which could have far-reaching effects.

**Comment by hendrikweisser** [comment]
Despite the issue with the pyro-carbamidomethyl, I think it's better to merge these changes than to leave them out, as they are an improvement over the previous state.

**Comment by hroest** [comment]
:+1: as this seems to improve the previous state but it is good to have a few tests cases on which somebody could work and improve the situation

File: src/openms/source/FORMAT/PepXMLFile.cpp [hunk 2]
```diff
@@ -69,7 +69,7 @@ namespace OpenMS
     hydrogen_ = *db->getElement("Hydrogen");
   }
 
-  const double PepXMLFile::mod_tol_ = 0.0001;
+  const double PepXMLFile::mod_tol_ = 0.001;
   const double PepXMLFile::xtandem_artificial_mod_tol_ = 0.0005; // according to cpp in some old version of xtandem somehow very small fixed modification (electron mass?) gets annotated by X!Tandem. Don't add them as they interfere with other modifications.
```
Comment (mwalzer) [inline src/openms/source/FORMAT/PepXMLFile.cpp]: I can vaguely remember that we had a discussion about that magic constant, any reason why the change?

Comment (hendrikweisser) [inline src/openms/source/FORMAT/PepXMLFile.cpp]: Yes, simply because the stricter tolerance didn't work for the data that I had.

**Comment by hendrikweisser** [comment]
Based on Hannes' approval, I've merged this.

**Comment by timosachsenberg** [comment]
ok. I hope changing the tolerance has no further implications. I think I recall that this formerly introduced some ambiguities but these might be resolved by now.

**Comment by hendrikweisser** [comment]
> I hope changing the tolerance has no further implications.

At least we have some time until the next release to fix any problems that might occur. ;-)

**Pull request merged**